### PR TITLE
P-tuning refactor Part 3/N 

### PR DESCRIPTION
**Pull request by arendu**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/dialogue/dialogue_gpt_classification_model.py [hunk 1]
```diff
@@ -291,7 +291,7 @@ def get_prompt_token_labels_for_megatron_gpt(self, input_ids, num_prompt_tokens)
 
     def get_virtual_prompt_ids_for_megatron_gpt(self, input_ids):
         if (
-            self.cfg.virtual_prompt_style == VirtualPromptStyle.PROMPT_TUNING
+            self.cfg.virtual_prompt_style == VirtualPromptStyle.P_TUNING
```
Comment (yidong72) [inline nemo/collections/nlp/models/dialogue/dialogue_gpt_classification_model.py:294]: `VirtualPromptStyle.PROMPT_TUNING` is removed from the enum?

File: nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py [hunk 2]
```diff
@@ -211,68 +170,25 @@ def init_prompt_encoder(self):
...
 
-        encoder_type = PromptEncoderType(self.cfg.p_tuning.get("encoder_type", "mlp").lower())
+        encoder_type = PromptEncoderType(self.cfg.p_tuning.get("encoder_type", "tpmlp").lower())
         self.prompt_encoder = PromptEncoder(
-            encoder_type=PromptEncoderType(encoder_type),
+            encoder_type=encoder_type,
             total_virtual_tokens=total_virtual_tokens,
             token_dim=self.hidden_size,
-            hidden_size=self.cfg.p_tuning.get("encoder_hidden", 2048),
+            hidden_size=self.cfg.p_tuning.get("encoder_hidden", self.hidden_size // 2),
```
Comment (yidong72) [inline nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:178]: is `self.hidden_size // 2` a good number to use?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py [hunk 3]
```diff
@@ -62,7 +64,7 @@
 __all__ = ['MegatronGPTPromptLearningModel']
 
 
-class MegatronGPTPromptLearningModel(MegatronBaseModel, TextGeneration):
+class MegatronGPTPromptLearningModel(MegatronBasePromptLearningModel):
```
Comment (yidong72) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:67]: why drop the `TextGeneration` interface?

File: nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py [hunk 4]
```diff
@@ -59,7 +58,9 @@ class MegatronBasePromptLearningModel(MegatronBaseModel, TextGeneration):
 
     def __init__(self, cfg: DictConfig, trainer: Trainer):
         super().__init__(cfg, trainer)
+        self.init_model(cfg, trainer)
```
Comment (Zhilin123) [inline nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:61]: I see here (in the lines above) that you have imports you don't use - maybe remove them in the refactor 4/n ? vscode should be able to identify them

File: nemo/collections/nlp/models/dialogue/dialogue_gpt_classification_model.py (hunk 1 above)
Comment (yidong72) [reply nemo/collections/nlp/models/dialogue/dialogue_gpt_classification_model.py:294]: okay. saw it. NVM

File: nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py [hunk 5]
```diff
@@ -82,37 +83,18 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
...
-        elif self.virtual_prompt_style == VirtualPromptStyle.P_TUNING:
+        if self.virtual_prompt_style == VirtualPromptStyle.P_TUNING:
             self.virtual_prompt_source = VirtualPromptSource.PROMPT_ENCODER
         elif self.virtual_prompt_style == VirtualPromptStyle.NO_PROMPT:
             self.virtual_prompt_source = VirtualPromptSource.NO_PROMPT
         else:
-            raise ValueError(
-                f"\nvirtual prompt style '{cfg.virtual_prompt_style}' not recognized, please use one of 'prompt-tuning' or 'p-tuning'"
-            )
+            raise ValueError(f"\nvirtual prompt style '{cfg.virtual_prompt_style}'")
```
Comment (Zhilin123) [inline nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:97]: Please indicate what they should use instead. e.g please use p-tuning

File: nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py [hunk 6]
```diff
@@ -211,68 +170,25 @@ def init_prompt_encoder(self):
         new_task = self.new_tasks[0]
         total_virtual_tokens = self.task_templates[new_task]["total_virtual_tokens"]
 
-        encoder_type = PromptEncoderType(self.cfg.p_tuning.get("encoder_type", "mlp").lower())
+        encoder_type = PromptEncoderType(self.cfg.p_tuning.get("encoder_type", "tpmlp").lower())
```
Comment (Zhilin123) [inline nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:173]: Is there a reason why default was changed?

File: nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py [hunk 7]
```diff
@@ -296,26 +212,33 @@ def load_state_dict(self, state_dict, strict: bool = True):
...
-
-            if (
-                self._prompt_encoder_key in state_dict
-                and self.virtual_prompt_source == VirtualPromptSource.PROMPT_ENCODER
-            ):
-                state_dict_ = state_dict[self._prompt_encoder_key]
-                self.prompt_encoder.load_state_dict(state_dict_, strict)
-
-    def embed_input_train(self, input_ids: Tensor, taskname_ids: Tensor):
+                raise ValueError("invalid virtual prompt source")
```
Comment (Zhilin123) [inline nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:220]: please indicate what is the correct config. E.g. please set this cfg.... to p_tuning

File: nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py (hunk 4 above)
Comment (Zhilin123) [reply nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:61]: There's quite a few in other files as well

**Review by Zhilin123 — APPROVED** [review]
LGTM, Great to see a lot of redundant code removed from this PR. Minor stylistic changes, but ok to merge, if you can address them in the next PR

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py (hunk 3 above)
Comment (arendu) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:67]: The base model containers it now!

File: nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py (hunk 6 above)
Comment (arendu) [reply nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:173]: No reason, it was during my copy based from the child model.

File: nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py (hunk 4 above)
Comment (arendu) [reply nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:61]: Agreed! will do!

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py (hunk 3 above)
Comment (arendu) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:67]: The base class now has TextGenetation.

File: nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py (hunk 2 above)
Comment (arendu) [reply nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:178]: It's a better default, for when the base model has different hidden sizes.

**Pull request merged**